home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ContactServiceImpl.java

```diff
@@ -0,0 +1,104 @@
+        User user = mapper.convert(userService.getUserById(userId), User.class);
+        contact.setUser(user);
+        contactRepository.save(contact);
+        return mapper.convert(contact, ContactDto.class);
+    }
+
+    @Transactional
+    @Override
+    public ContactDto updateContact(Long id, ContactDto updateContactDto) {
+        Optional<Contact> contactOptional = contactRepository.findById(id);
+        if (contactOptional.isPresent()) {
+            Contact contact = contactOptional.get();
+            ContactDto contactDto = mapper.convert(contact, ContactDto.class);
+            if (contactDto.getContactType().equals(updateContactDto.getContactType())) {
+                return checkContactType(contact, updateContactDto);
+            } else {
+                throw new IllegalArgumentException("Type of the contact doesn't match");
+            }
+        } else {
+            throw new NotFoundHomeException("User with id:" + id + " is not found");
+        }
+    }
+
+    private ContactDto checkContactType(Contact contact, ContactDto updateContactDto) {
+        if (contact.getContactType().equals(ContactType.PHONE)) {
+            return updatePhone((Phone) contact, (PhoneContactDto) updateContactDto);
+        } else if (contact.getContactType().equals(ContactType.EMAIL)) {
+            return updateEmail((Email) contact, (EmailContactDto) updateContactDto);
+        } else {
+            throw new IllegalArgumentException("Invalid type of the contact");
+        }
+    }
+
+    private ContactDto updatePhone(Phone phone, PhoneContactDto phoneContactDto) {
+        phone.setPhone(phoneContactDto.getPhone());
+        contactRepository.save(phone);
+        return mapper.convert(phone, PhoneContactDto.class);
+    }
+
+    private ContactDto updateEmail(Email email, EmailContactDto emailContactDto) {
+        email.setEmail(emailContactDto.getEmail());
+        contactRepository.save(email);
+        return mapper.convert(email, EmailContactDto.class);
+    }
+
+    @Transactional
+    @Override
+    public Page<ContactDto> getAllContacts(Integer pageNumber, Integer pageSize,
+                                            Specification<Contact> specification) {
+        return contactRepository.findAll(specification, PageRequest.of(pageNumber - 1, pageSize))
+                .map(contact -> mapper.convert(contact, ContactDto.class));
+    }
+
+    @Override
+    public ContactDto getContactById(Long id) {
+        Contact contactResponse = contactRepository.findById(id)
+                .orElseThrow(() -> new NotFoundHomeException("Can't find contact with given ID:" + id));
+        return mapper.convert(contactResponse, ContactDto.class);
+    }
+
+    @Override
+    public void deactivateContact(Long id) {
+        contactRepository.findById(id)
+                .orElseThrow(() -> new NotFoundHomeException("Can't find contact with given ID:" + id));
+        contactRepository.deleteById(id);
```

Comment:
Do we really want completely delete it? As we decided previously we should follow enabled/disabled approach so we will be able to keep historical information.

Reply:
Yeah, sure. Done.